drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+    PTDI_CONNECTION_INFORMATION ConnectionInformation = NULL;
+    NTSTATUS status;
+
+    TargetAddress = TdiTransportAddressFromSocketAddress(SocketAddress);
+    if (TargetAddress == NULL)
+        return NULL;
+
+    status = TdiBuildConnectionInfo(&ConnectionInformation, TargetAddress);
+    ExFreePoolWithTag(TargetAddress, 'ADDR');
+
+    if (!NT_SUCCESS(status))
+    {
+        if (ConnectionInformation != NULL)
+        {
+            ExFreePoolWithTag(ConnectionInformation, TAG_AFD_TDI_CONNECTION_INFORMATION);
+        }
+        return NULL;
+    }
+    return ConnectionInformation;
+}
+
+static NTSTATUS WSKAPI
+WskBind(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR LocalAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    NTSTATUS status;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (s->LocalAddressHandle != INVALID_HANDLE_VALUE)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = INVALID_HANDLE_VALUE;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+    Irp->IoStatus.Status = status;
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
```

Comment:
You need to free this in the failure case / completion routine.

Reply:
Ok I reworked that and implemented it using "chrismas tree" (don't know how the generic
name for this is):

    if (allocate r1 fails)
       goto out

    if (allocate r2 fails)
        goto out_free_r1

    if (allocate r3 fails)
        goto out_free_r2
    do_something()
    return success

    out_free_r2:
        free r2

    out_free_r1:
        free r1;

    out:
       return error

There were more instances of unfreed resources if a later allocation fails.
All of these should be fixed now (if not please let me know).

Reply:
This is fine, though it's often simpler to just have a single Exit label and signify whether a given variable needs to be freed by whether it is non-null.

Reply:
You are right but let's keep it like this for now.